internal/checkcontroller/reconcile.go

```diff
@@ -0,0 +1,518 @@
+		}
+		return nil
+	}
+
+	logger.V(1).Info("Creating rebooter pod")
+	if err := r.CreateRebooterPod(ctx, workerName, nodeName, namespace); err != nil {
+		return fmt.Errorf("failed to create rebooter pod for node %s: %w", nodeName, err)
+	}
+
+	logger.V(1).Info("Rebooter pod created successfully")
+	return nil
+}
+
+// GetRebooterPod gets the rebooter pod for the node with the given name in the given namespace.
+func (r *CheckControllerReconciler) GetRebooterPod(ctx context.Context, nodeName, namespace string) (*corev1.Pod, error) {
+	pod := &corev1.Pod{}
+	err := r.Get(ctx, client.ObjectKey{Name: fmt.Sprintf("node-rebooter-%s", nodeName), Namespace: namespace}, pod)
+	if err != nil {
+		return nil, fmt.Errorf("error getting rebooter pod: %w", err)
+	}
+	return pod, nil
+}
+
+// SetNodeConditionIfNotExists sets a custom condition on the node with the given name if it does not already exist.
+func (r *CheckControllerReconciler) SetNodeConditionIfNotExists(
+	ctx context.Context,
+	node *corev1.Node,
+	conditionType corev1.NodeConditionType,
+	status corev1.ConditionStatus,
+	message string,
+) error {
+	logger := log.FromContext(ctx).WithName("SetNodeConditionIfNotExists").WithValues("nodeName", node.Name)
+
+	// The field node.Status.Conditions belongs to the status of the Node resource.
+	// In Kubernetes, the status is considered a "system-owned" object and cannot be
+	// modified using a regular Update call.
+	// Instead, changes to the status must be made using the Status().Update method.
+	for i, cond := range node.Status.Conditions {
+		if cond.Type == conditionType {
+			if cond.Status == status {
+				logger.V(1).Info(fmt.Sprintf("Node already has condition %s set to %s", conditionType, status))
+				return nil
+			}
+			logger.V(1).Info("Updating existing condition to %s", status)
+			node.Status.Conditions[i] = corev1.NodeCondition{
+				Type:    conditionType,
+				Status:  status,
+				Message: message,
+				LastTransitionTime: metav1.Time{
+					Time: time.Now(),
+				},
+				LastHeartbeatTime: metav1.Time{
+					Time: time.Now(),
+				},
+			}
+			return r.Status().Update(ctx, node)
+		}
+	}
+
+	logger.V(1).Info("Adding new condition to node")
+	node.Status.Conditions = append(node.Status.Conditions, corev1.NodeCondition{
+		Type:    conditionType,
+		Status:  status,
+		Message: message,
+		LastTransitionTime: metav1.Time{
+			Time: time.Now(),
+		},
+		LastHeartbeatTime: metav1.Time{
+			Time: time.Now(),
+		},
+	})
+	return r.Status().Update(ctx, node)
+}
+
+// GetPodNodeName gets the node name of the pod with the given name in the given namespace.
+func (r *CheckControllerReconciler) GetPodNodeName(ctx context.Context, nodeName, namespace string) (string, error) {
+	pod := &corev1.Pod{}
+	err := r.Get(ctx, client.ObjectKey{Name: nodeName, Namespace: namespace}, pod)
+	if err != nil {
+		return "", fmt.Errorf("error getting pod: %w", err)
+	}
+	if pod.Spec.NodeName == "" {
+		return "", fmt.Errorf("pod has no node name")
```

Comment:
for future debugging:
```suggestion
		return "", fmt.Errorf("pod %s has no node name", podName)
```